Userland/Services/SystemServer/main.cpp

```diff
@@ -155,9 +174,64 @@ static ErrorOr<void> spawn_device_mapper_process()
     return {};
 }
 
+static ErrorOr<void> prepare_tmpfs_system_devicemap_directory()
+{
+    dbgln("Creating /tmp/system/devicemap directory");
+
+    TRY(Core::System::mkdir("/tmp/system/"sv, 0755));
+    TRY(Core::System::mkdir("/tmp/system/devicemap/"sv, 0755));
+    TRY(Core::System::mount(-1, "/tmp/system/devicemap/"sv, "ram"sv, MS_NOEXEC | MS_NOSUID | MS_NODEV));
+    TRY(Core::System::mkdir("/tmp/system/devicemap/nodes/"sv, 0755));
+    TRY(Core::System::mkdir("/tmp/system/devicemap/nodes/block/"sv, 0755));
+    TRY(Core::System::mkdir("/tmp/system/devicemap/nodes/char/"sv, 0755));
+
+    TRY(Core::System::mkdir("/tmp/system/devicemap/family/"sv, 0755));
+
+    auto build_tmp_system_devicemap_directory = [](StringView node_type_directory, JsonObject const& major_number_allocation_object) -> ErrorOr<void> {
+        auto allocated_number = major_number_allocation_object.get_u64("allocated_number"sv).value_or(0);
+        auto path = LexicalPath("/tmp/system/devicemap/nodes/").append(node_type_directory).append(MUST(String::number(allocated_number)));
+        TRY(Core::System::mkdir(path.string(), 0755));
+
+        auto possible_family_name = major_number_allocation_object.get_byte_string("family_name"sv);
+        if (!possible_family_name.has_value())
+            return Error::from_string_literal("Major allocation has no valid family name");
+
+        auto symlink_path = LexicalPath("/tmp/system/devicemap/family/").append(possible_family_name.value());
+        TRY(Core::System::symlink(path.string(), symlink_path.string()));
+        return {};
+    };
+
+    {
+        auto file = TRY(Core::File::open("/sys/kernel/chardev_major_allocs"sv, Core::File::OpenMode::Read));
```

Comment:
Sorry, I don't know the kernel super well and am confused by the code in this commit. From why understand, previously we set up a procfs that exports the block and character device major allocs from the kernel as json `/sys/kernel/chardev_major_allocs` (…or blockdev). This here then reads that file and sets up a mapping from user space device numbers to kernel device numbers (? Is that right? Previously, the user space and kernel space numbers matched, and it looks like they still mostly do? Userspace numbers are around line 150 in this file, and kernel space numbers in that new central header file you're adding in this PR?)

How often does this code here run? Just once, right?

And then DeviceMapper already runs once when a device is added and once when it's removed, and that's what adds and removes the hotplug files, yes?

Reply:
This is not `procfs`, but a `sysfs` node :)
Also, yes, this `sysfs` file exports major numbers for character devices. It runs once in the `SystemServer` code and the general idea is to ensure that we never hardcode the allocations in the kernel and userspace (so it is 2 sources of truth) but only have one source of truth on allocations of numbers.

`SystemServer` ensures that we have a directory in `/tmp` before it spawns `DeviceMapper` as a daemon.
Then `DeviceMapper` can use to spawn files there so other programs can watch for such events.
For example, if you insert a mouse or a keyboard, `WindowServer` will watch on `/tmp/system/devicemap/family/mouse/` and `/tmp/system/devicemap/family/mouse/` directories, and when `DeviceMapper` will get a notification for such event from the kernel, it will spawn files in these directories and `WindowServer` will see these events :)

So it costs a bit of time that we spend on running this logic, but it spares us from having 2 identical hardcoded lists in userspace and kernel. In the future, if we want, we can use this to insert more allocations in runtime when a user inserts loadable kernel modules.

Reply:
> the general idea is to ensure that we never hardcode the allocations in the kernel and userspace (so it is 2 sources of truth) but only have one source of truth on allocations of numbers.

Which numbers do you mean here? E.g. `226` for `CharacterDeviceFamily::GPU`? If so, that's still in both Kernel/Devices/MajorNumberAllocation.h and in Userland/Services/DeviceMapper/DeviceEventLoop.cpp. Do you mean different numbers? Or is what you've written aspirational and not done yet?

(Either's fine, just want to understand things. The rest of your reply makes sense to me, I just don't understand this one part yet.)

Reply:
> Which numbers do you mean here? E.g. 226 for CharacterDeviceFamily::GPU? If so, that's still in both Kernel/Devices/MajorNumberAllocation.h and in Userland/Services/DeviceMapper/DeviceEventLoop.cpp. Do you mean different numbers? Or is what you've written aspirational and not done yet?

This is aspirational, as I do intend to eventually get rid of that list in userspace or at least narrow it down to hard-coded device family names and not numbers :)